### ISSUE TITLE
About collaboration

### ISSUE THREAD

**Issue by waj** [crystal-lang/crystal#1805]
I extracted some off topic messages from #1710 so I can reply here without continue mixing with the original topic.

@technorama commented on Oct 21, 2015, 7:44 PM GMT

> @waj You've already expressed that not only my ideas but my code are not welcome regardless of what they are, even if it's your design (that you don't have or won't share).  You're unwilling to collaborate on the design online or if I fly to Argentina (your excuse was a lie).  You left me waiting for any communication whatsoever for over a week when you had already told another member you wouldn't accept my offer.  Later you claim threading is fun, not a priority, don't have the time, your contributions aren't welcome no matter what lengths you go to (even temporarily moving to Argentina and learning Spanish).
>
> Now you're going to import an external project to one that doesn't want what I can do?  Especially when I have extensive OpenSSL experience?
>
> @datanoise accepted my PR's quickly.  The same has not been true of crystal.
>
> Your idea of "enjoy" is very different from mine.

Seems like you're inferring that I have something personal against you. Well, I do have something, which is the constant unpleasant way that you have to express to others. That could be something belonging to your personality, but that's all. I respect your ideas and the work you done so far, and I think you're a respectable software developer.

Regarding the particular topic that you're referring about design collaboration (multithreading in Crystal) all I did was try to let you know that I found really hard to process your ideas into our vision of Crystal at this point. You were working hard into GC and threadind issues but just to find that we are still not so sure about the direction we want to take in that area. So I thought it could be a good idea to try to stop you from working on that topic until we have a plan. Seems like this was actually a bad idea because of course you're free to even fork the project and have your own direction if you want. Still I think we're free to accept or reject any work done by others no matter how much time was spent, but I wanted to minimize that posibility.

And no, I don't want to meet you in person to collaborte on this. I think I have the right to not give a logical explanation.

@mverzilli commented on Oct 21, 2015, 9:10 PM GMT

> @technorama well if you did come to Argentina and met @waj and @asterite you'd realise they're two of the nicest guys working on software out there :).

> I think you're taking things way too personal: "not only *my ideas* but *my code* are not welcome", "*you left me* waiting for a week", "import an external project to one that doesn't want what *I can do*", etc.

> Please realise that everyone here is contributing to the extent each one can, and @waj and @asterite are no exception. If you didn't receive a reply for a week it's because there's a lot going on: work commitments, personal lives and lots of other aspects of Crystal itself.

> Everyone here is giving you credit for some important contributions you already made, but some people is asking you for a little bit more kindness on your interactions.

> I'm sure your OpenSSL experience and the rest of your ideas are being taken into account just as anyone else's, so don't let your manners cast a shadow over that.

@technorama commented on Oct 21, 2015, 9:47 PM GMT

> @mverzilli My issues are not with the time delays but with the way it was handled.  Another project member was told about rejecting my offer within a day or 2 and the reasons why.  It took over a week to tell me directly.   It's obviously not a time issue if they have time to speak to others but not tell me directly, ask me questions or answer any of mine. Many attempts were made to contact asterite and waj through irc, email, issues, and wiki entries.  No comments were ever made on my proposals except "don't work on that".  Every collaboration effort was made on my part and none on theirs except silence and a belated "no".

> Maybe they are nice guys in person but their communication sucks.

Maybe you're right and my communication suck. I'll try to improve that. Thanks for the heads up.

@technorama commented on Oct 21, 2015, 9:56 PM GMT

> I've bent over backwards to collaborate on threading, get a spec approved, or get anything at all.  What will be accepted?  Who can tell me that?  Why should I create PR's when I don't know if they will be approved?  Why create issues and email proposed designs if they are met with silence?  What other options do I have other than complaining publicly or forking?

> This isn't just about threading.  It's the smallest of issues sitting around for months without a response: #1284.

Again, that's why I tried to warn you asap about the posibility that we wanted to take a different direction and I didn't want you were working in vain. I think my mistake here was that I did it in private. From now on, all aspects of communication from me about Crystal will be done in public (IRC, GitHub or through the mailing list)

@kirbyfan64 commented on Oct 21, 2015, 10:14 PM GMT

> @technorama BUT you're missing the point. This is a project managed by *people*. Which means, there inevitably will be delays, accidental non-replies (you know, when you read a GitHub issue, say you'll reply later, and don't), and stuff like that. @datanoise's project is small and has WAY fewer issues open than Crystal.

> Also, about the PR approval: normally, it's best to open an issue related to whatever and mention that you'll implement it...but don't implement it yet. That way, if it gets closed, there's no lost work, and you can avoid constantly tweaking PR commits over and over again.

@technorama commented on Oct 21, 2015, 11:04 PM GMT

> @kirbyfan64 You're missing the point.  I've tried all of those things.  First I created PR's which sat idle.  Then I tried issues describing the proposed interfaces which also sat idle.  Sometimes if I create the issue (which is ignored) and then submit a PR the code gets merged.  Other times neither get any feedback.  Either way I'm wasting my time.

> It's requires me reading their minds to know what will or won't be approved.

> Let me say this again.  As *people* their communication about PR's and issues suck.

I'm so sorry you feel you're wasting time. Lots of stuff contributed by you was merged already and is currently running on every program compiled with Crystal. If an issue or PR is sitting idle without answer, please try bumping the thread again. Maybe sometimes we still didn't have the time to reply. Maybe we still are not sure what to say. Or maybe we're just humans and forgot to reply.

Put yourself in our position for a moment and think about a scenario with the opposite situation: we decide to delegate some important area of the project to someone but we get no answer after some time or the work is not what we expected. Should we merge it anyway? Are either us or that person wasting time?

@technorama commented on Oct 21, 2015, 11:08 PM GMT

> @kirbyfan64 You're point about @datanoise's project only supports my position.  Crystal moves too slow with issues and PR's.   The bottlenecks are @asterite and @waj.  They're busy people and that's ok but funneling everything through them is only going to make the problem worse.  I thought distributed development was a solution for this problem but apparently I'm mistaken.

Crystal moves at the speed and pace we want it to move. No rush, no pressure, we want to take the time it takes. How could we be bottlenecks? We created this project. Maybe you have a misconception about how open source should work.

My honest desire is not to extend this to an endless conversation. Instead I wish you reconsider the position you want to take and keep belonging to this community because I still consider you a valuable member.

**Comment by vyp** [comment]
> I thought distributed development was a solution for this problem but apparently I'm mistaken.

To add to the original project, only those with write access to the original project can do it (by accepting others' patches), how else could it be done? Distributed development cannot bypass that. Remember it is possible as always with free software to make a fork/derivative (although yes, forking a big project like this, especially a language, is probably practically infeasible, at least without a lot of resources).

**Comment by asterite** [comment]
I want to add a couple of things to what @waj said.

@technorama Imagine this scenario: we never create Crystal. Then you come to us and say: "hey, you, why didn't you create Crystal? You are being a bottleneck, I want to help you build the language you always had in mind but you are not letting me do that". As a complaint.

Now back to the real scenario: we create Crystal after spending about 3 years to do so, with all of the efforts being spent in our free time. Yes, we have a bountysource program but that doesn't cover the costs at all. We have no client, no one is really paying us or asking us to do anything. We go at our pace, because we don't like to rush things. We love this project and have a lot of fun doing it. And at the same time we might help bring the world a valuable tool, and something fun to use. We like to sit down or be in front of a blackboard and discuss things in Spanish. Of course you can learn Spanish, but it will take you months to learn it to a level where we can comfortably discuss these things face to face. And even if you come here, as I already told you, we don't spend all of our time in Crystal. We have work to do, we have family, cats, and then we spend some time doing Crystal. Some of that time we have to answer issues, review pull requests, reply messages in google groups, maybe reply some things on IRC. The rest of that time (which now you can realize is pretty little) we can focus on continuing designing the language. And we wouldn't like to delegate the design of many core things to somebody else.

If the above seems unfair from our side, think of other languages. Did you, or anyone else, had a chance to define the design of Java, C# or Go? No. They have been defined by a core team without opinions or help from anyone else. I imagine Matz says "no" more than "yes" to many ideas and contributions from others.

So we could just make Crystal closed source, or just make the pull requests button go away. We don't. We provide it so others can help us build the language and propose new ideas and features, but we reserve the right to decide what we think is good or bad. With time, our opinions might change and some wishes might happen (like the formatter, or splat type restriction, or even some of your pull requests). Many other followers got rejected PR and ideas from us, and most accept our design choices, our pace and our efforts.

@vyp Forking this language is not far fetched at all. Remember: we are just a few doing it, and we do it in our free time. It seems @technorama has much more time than us, so forking it and continuing it might deliver a better, more efficient and correct language in less time. If that happens, at that point we can decide to merge the changes or just continue using the other language. See also the [onyx-lang](https://github.com/ozra/onyx-lang).

**Comment by vyp** [comment]
Well I meant in the sense of gathering a community, having website, docs etc. But I haven't seen Onyx before, that's really cool @ozra, but I wish you had some samples of the changes in syntax somewhere! https://github.com/ozra/onyx-lang#what-does-the-current-syntax-draft-look-like :stuck_out_tongue:

**Comment by ozra** [comment]
@technorama - You seem like a real sharp developer, intense! Now, a sidenote: I grew up with a alcoholic dad, there was lots of yelling, abusive language and fist fighting - I've learned to turn anything thrown at me to a summer breeze (ok, somewhat). But most people don't. Just as we like to format code so it fits a project, we can benefit a lot from formatting our language to communicate well in a project. 

Your ideas deserve some public relations polishing - good arguments, and your skills are enough - if you try to tone down you equally intense language and keep the intensity to the coding, I think those who are more sensitive to choice of words will be more willing to listen, read, understand and ultimately merge. 

People skills count when dealing with people. Keep up the good work! I really like your ideas!

@asterite @vyp - What a surprise to see a mention on Onyx here :-) Yes I reckon a lot more people than me, are seeking a "next gen" language to use, and not a Rubyish language. Now Crystal happen to be a really interesting next gen language at its core - I'm really impressed by it!
My intention is to get a language where everything's up to debate via RFC's to form an as educated idea as possible about the best way forward. And to stay _completely_ compatible with Crystal. Because many people do love ruby syntax - and I see no reason to not share a module universe and be a "family". This way, if it hopefully takes off by interest, there are two pathways to the same core yielding PR's, fixes to libs, etc.
I realize that some people might frown upon it. But my intentions are all good :)

@vyp - I have more work that is not pushed to origin yet. I'll make sure to get some samples up. 

The motto I go by initially is to support all the ideas _I_ have so far found to be interesting or think might be interesting to others (I already have some favourites, but figure it's better to get input, hopefully a superior idea grows from collaboration), in parallel, so that it can later be distilled down. 

You can basically write _almost_ ruby syntax, _almost_ python syntax, a more haskellish and livescriptish slant - all mixed as you will. There are no ambiguities the way it's designed now - but obviously it would be better to reduce it to not lock up other syntactic sugars that could become ambiguous with this liberal RFC-stage setup.

One of the important concepts, structurally, is that it uses indent sensitivity (keep good form!), but one can also use end-keywords to assert and ensure structure (unlike pythons optional end-keywords which are just pre-processed and not interchangeably used.). A kind of double life guard. What you see is much more likely what you intended and also get.

The idea about allowing diversity is that different kinds of code may benefit from different choices of style, and also that different users like it differently. I'd like to get it so that you have a formatting-conf for checking in (project style), and another for checking out (your style). The ideas awoke from @technorama 's issue on elastic tabstops in #1682, and the fact that I myself have used a crude search n replace back/forth method for editing others code.

I also have a lot of RFC's prepared that I will post in the issues, once I've gotten them concise and divided enough. So that there's a base structure for initial commentry.

_It could likely also work as an experimental ground for lib and core-features, that might be interesting for Crystal to merge in, and otherwise those features might drop out of Onyx instead, to maintain the long term goal of full co-compile compatibility_.

And, well same thing as for Crystal goes for me, I have wood to cut for the winter, cat's, dog and lady of the house to attend to ;-)

_**Sorry for taking up valuable screen estate about Onyx here, @asterite, @waj, etc!!**_

**Comment by technorama** [comment]
@waj What everyone sees from me is months of frustration building.  I tried contacting @waj, @asterite, @jhass in public (issues, PR's, irc) and private.  So far @jhass has been wonderful.  He's fair, diplomatic, merges or comments quickly,  and is helpful even in responding with IDK.  With @waj and @asterite it's mostly 1 sided conversations.  Private messages, channel messages go mostly unanswered.  Emails have never been answered.  

My tone was completely different months ago.
> 2015-07-22 10:01:53     dzv     thank you for looking but there is a bigger issue
> 2015-07-22 10:02:20     dzv     how can we grow the community if everything is on your shoulders?
> 2015-07-22 10:02:41     dzv     perhaps jhass could serve as a merge approver?
> 2015-07-22 10:03:07     dzv     or some other additional people?
> 2015-07-22 10:13:02     waj     eventually there will be more maintainers
> 2015-07-22 10:13:33     waj     but I think I prefer to keep control until we reach a more stable state
> 2015-07-22 10:14:19     dzv     there are bugfixes and other contributions i would like to make
> 2015-07-22 10:14:45     dzv     but i'm not going to commit myself to them if my commits won't be accepted
> 2015-07-22 10:14:54     dzv     and right now there's no feedback.  i have no idea if i'm wasting my time or not
> 2015-07-22 10:15:32     dzv     maybe there is a middle ground
> 2015-07-22 10:15:48     waj     if you look at the commit history you?ll see there are many PR accepted every day  
> 2015-07-22 10:16:27     dzv     expanding community contributions from trusted members.  i'm not asking for myself to be one of them.  i'm asking if you would consider promoting some other members.  perhaps there are some active members already
> 2015-07-22 10:16:54     dzv     i watch the commits every day but see nothing about my own which makes me think my approach is wrong
> 2015-07-22 10:17:33     dzv     or that you only want bug fixes.  or secretly hate Process#spawn which solves a different problem than Process#run
> 2015-07-22 10:18:45     waj     no? it?s nothing like that. I?m sorry you get the wrong impression. We?re not full time commited to the project anyway. We need to work also so we can pay our bills :)
> 2015-07-22 10:19:22     dzv     understood.  i would like crystal to gain enough traction where i can pay my bills by writing programs in it

It's not like I'm raising a new issue.  This issue is just becoming public now because I don't see improvement (with the exception of promoting @jhass).  Adding @jhass as a contributor was not enough.  He's not empowered to make language design changes and often punts to @waj or @asterite.

If @waj and @asterite followed @jhass's public relations lead there wouldn't be a problem.

**Comment by asterite** [comment]
I'm closing this as there's nothing to track here.
